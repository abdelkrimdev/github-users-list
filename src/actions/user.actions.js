import { userConstants } from '../constants/user.constants';
import { getAllUsers, getSingleUser } from '../services/github';

export const userActions = {
    getAll: (since = 0) => {
        return async (dispatch) => {
            try {
                dispatch(request());
                
                const payload = await getAllUsers(since);
                
                dispatch(success(payload));
            } catch (error) {
                dispatch(failure(error));
            }
        };

        function request() {
            return { type: userConstants.USER_ALL_REQUEST };
        }
        function success(users) {
            return { type: userConstants.USER_ALL_SUCCESS, users };
        }
        function failure(error) {
            return { type: userConstants.USER_ALL_FAILURE, error };
        }
    },
    getByUsername: (username) => {
        return async (dispatch) => {
            try {
                dispatch(request());
    
                const payload = await getSingleUser(username);
                
                dispatch(success(payload));
            } catch (error) {
                dispatch(failure(error));
            }
        };

        function request() {
            return { type: userConstants.USER_SPECIFIC_REQUEST };
        }
        function success(user) {
            return { type: userConstants.USER_SPECIFIC_SUCCESS, user };
        }
        function failure(error) {
            return { type: userConstants.USER_SPECIFIC_FAILURE, error };
        }
    }
};
