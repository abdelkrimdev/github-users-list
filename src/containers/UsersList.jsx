import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import { Alert, Avatar, BackTop, Button, List, Spin } from 'antd';

import { userActions } from '../actions/user.actions';

export class UsersList extends Component {
    static propTypes = {
        requesting: PropTypes.bool,
        users: PropTypes.array,
        error: PropTypes.object,
        loadUsers: PropTypes.func.isRequired,
        unloadUsers: PropTypes.func.isRequired
    }

    state = {
      loadingMore: false,
      showLoadingMore: true
    }

    componentDidMount() {
        this.props.loadUsers();
    }

    onLoadMore = () => {
        const latestUser = this.props.users.slice(-1)[0];
        this.props.loadUsers(latestUser.id);
    }

    componentWillUnmount() {
        this.props.unloadUsers();
    }

    render() {
        const { requesting, users, error } = this.props;
        const { loadingMore, showLoadingMore } = this.state;

        return (
            <div id="users-list">
                { error && <Alert message="Error" description={ error.message } type="error" showIcon closable /> }
                <div className="title-bar">
                    <h1>GitHub's Users List</h1>
                </div>
                <BackTop />
                <List
                    loading={ requesting }
                    size="large"
                    itemLayout="horizontal"
                    grid={{ gutter: 16, xs: 1, sm: 1, md: 2, lg: 2, xl: 4, xxl: 4 }}
                    loadMore={ 
                        showLoadingMore ? (
                            <div className="show-more">
                                { loadingMore && <Spin /> }
                                { !loadingMore && <Button onClick={ this.onLoadMore }>loading more</Button> }
                            </div>
                        ) : null
                    }
                    dataSource={ users }
                    renderItem={ user => (
                        <List.Item actions={ [ <Link className="ant-btn" to={ `/user/${user.login}` }>details</Link> ] }>
                            <List.Item.Meta
                                avatar={ <Avatar size="large" src={ user.avatar_url } /> }
                                title={ user.login }
                                description={ user.type }
                            />
                        </List.Item>
                    ) }
                />
            </div>
        );
    }
}

const mapStateToProps = (state) => {
    const { requesting, users, error } = state.allUsers;
    return {
        requesting,
        users,
        error
    };
};

const mapDispatchToProps = (dispatch) => {
    return {
        loadUsers: (since = 0) => dispatch(userActions.getAll(since)),
        unloadUsers: () => dispatch(userActions.unloadAll())
    };
};

export default connect(mapStateToProps, mapDispatchToProps)(UsersList);
