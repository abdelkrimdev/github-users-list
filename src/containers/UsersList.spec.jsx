import React from 'react';
import { createMockStore } from 'redux-test-utils';
import { shallow } from 'enzyme';

import UsersList from './UsersList';
import { userActions } from '../actions/user.actions';

describe('Users List Component.', () => {
    let store;
    let wrapper;

    beforeEach(() => {
        store = createMockStore({
            allUsers: {}
        });
        wrapper = shallow(<UsersList store={ store } />).dive();
    });

    it('should be defined.', () => {
        expect(UsersList).toBeDefined();
    });

    it('should render correctly.', () => {
      expect(wrapper).toMatchSnapshot();
    });

    it('should dispatch load users action on user opens the page.', () => {
        const expectedActions = [
            userActions.getAll()
        ];
        expect(store.isActionDispatched(userActions.getAll())).toBeTruthy();
    });

    it('should render a title for the page.', () => {
        expect(wrapper.find('h1')).toHaveLength(1);
    });

    it('should render the BackTop component.', () => {
        expect(wrapper.find('BackTop')).toHaveLength(1);
    });

    it('should render a List component.', () => {
        expect(wrapper.find('List')).toHaveLength(1);
    });

    it('should unload users on user leaves the page.', () => {
        const expectedActions = [
            userActions.unloadAll()
        ];
        wrapper.unmount();
        expect(store.getActions()).toEqual(jasmine.arrayContaining(expectedActions));
    });
});
