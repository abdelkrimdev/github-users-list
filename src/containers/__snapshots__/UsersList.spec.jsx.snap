// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Users List Component. should render correctly. 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <UsersList
    error={undefined}
    loadUsers={[Function]}
    requesting={undefined}
    store={
        Object {
            "dispatch": [Function],
            "getAction": [Function],
            "getActions": [Function],
            "getState": [Function],
            "isActionDispatched": [Function],
            "isActionTypeDispatched": [Function],
            "subscribe": [Function],
          }
    }
    storeSubscription={
        Subscription {
            "listeners": Object {
              "clear": [Function],
              "get": [Function],
              "notify": [Function],
              "subscribe": [Function],
            },
            "onStateChange": [Function],
            "parentSub": undefined,
            "store": Object {
              "dispatch": [Function],
              "getAction": [Function],
              "getActions": [Function],
              "getState": [Function],
              "isActionDispatched": [Function],
              "isActionTypeDispatched": [Function],
              "subscribe": [Function],
            },
            "unsubscribe": undefined,
          }
    }
    unloadUsers={[Function]}
    users={undefined}
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        undefined,
        <div
          className="title-bar"
>
          <h1>
                    GitHub's Users List
          </h1>
</div>,
        <BackTop
          visibilityHeight={400}
/>,
        <List
          bordered={false}
          dataSource={Array []}
          grid={
                    Object {
                              "gutter": 16,
                              "lg": 2,
                              "md": 2,
                              "sm": 1,
                              "xl": 4,
                              "xs": 1,
                              "xxl": 4,
                            }
          }
          itemLayout="horizontal"
          loadMore={
                    <div
                              className="show-more"
                    >
                              <Button
                                        ghost={false}
                                        loading={false}
                                        onClick={[Function]}
                                        prefixCls="ant-btn"
                              >
                                        loading more
                              </Button>
                    </div>
          }
          loading={false}
          pagination={false}
          prefixCls="ant-list"
          renderItem={[Function]}
          size="large"
          split={true}
/>,
      ],
      "id": "users-list",
    },
    "ref": null,
    "rendered": Array [
      undefined,
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <h1>
            GitHub's Users List
</h1>,
          "className": "title-bar",
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "GitHub's Users List",
          },
          "ref": null,
          "rendered": "GitHub's Users List",
          "type": "h1",
        },
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "visibilityHeight": 400,
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "bordered": false,
          "dataSource": Array [],
          "grid": Object {
            "gutter": 16,
            "lg": 2,
            "md": 2,
            "sm": 1,
            "xl": 4,
            "xs": 1,
            "xxl": 4,
          },
          "itemLayout": "horizontal",
          "loadMore": <div
            className="show-more"
>
            <Button
                        ghost={false}
                        loading={false}
                        onClick={[Function]}
                        prefixCls="ant-btn"
            >
                        loading more
            </Button>
</div>,
          "loading": false,
          "pagination": false,
          "prefixCls": "ant-list",
          "renderItem": [Function],
          "size": "large",
          "split": true,
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          undefined,
          <div
            className="title-bar"
>
            <h1>
                        GitHub's Users List
            </h1>
</div>,
          <BackTop
            visibilityHeight={400}
/>,
          <List
            bordered={false}
            dataSource={Array []}
            grid={
                        Object {
                                    "gutter": 16,
                                    "lg": 2,
                                    "md": 2,
                                    "sm": 1,
                                    "xl": 4,
                                    "xs": 1,
                                    "xxl": 4,
                                  }
            }
            itemLayout="horizontal"
            loadMore={
                        <div
                                    className="show-more"
                        >
                                    <Button
                                                ghost={false}
                                                loading={false}
                                                onClick={[Function]}
                                                prefixCls="ant-btn"
                                    >
                                                loading more
                                    </Button>
                        </div>
            }
            loading={false}
            pagination={false}
            prefixCls="ant-list"
            renderItem={[Function]}
            size="large"
            split={true}
/>,
        ],
        "id": "users-list",
      },
      "ref": null,
      "rendered": Array [
        undefined,
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <h1>
              GitHub's Users List
</h1>,
            "className": "title-bar",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "GitHub's Users List",
            },
            "ref": null,
            "rendered": "GitHub's Users List",
            "type": "h1",
          },
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "visibilityHeight": 400,
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "bordered": false,
            "dataSource": Array [],
            "grid": Object {
              "gutter": 16,
              "lg": 2,
              "md": 2,
              "sm": 1,
              "xl": 4,
              "xs": 1,
              "xxl": 4,
            },
            "itemLayout": "horizontal",
            "loadMore": <div
              className="show-more"
>
              <Button
                            ghost={false}
                            loading={false}
                            onClick={[Function]}
                            prefixCls="ant-btn"
              >
                            loading more
              </Button>
</div>,
            "loading": false,
            "pagination": false,
            "prefixCls": "ant-list",
            "renderItem": [Function],
            "size": "large",
            "split": true,
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
